tests/test_observers/test_s3_observer.py

```diff
@@ -0,0 +1,229 @@
+    # twice, so for the FileStorageObserver to read it, we need to close it.
+    f = tempfile.NamedTemporaryFile(suffix='.py', delete=False)
+
+    f.content = 'import sacred\n'
+    f.write(f.content.encode())
+    f.flush()
+    f.seek(0)
+    f.md5sum = hashlib.md5(f.read()).hexdigest()
+
+    f.close()
+
+    yield f
+
+    os.remove(f.name)
+
+
+def _bucket_exists(bucket_name):
+    s3 = boto3.resource('s3')
+    try:
+        s3.meta.client.head_bucket(Bucket=bucket_name)
+    except ClientError as e:
+        if e.response['Error']['Code'] == '404':
+            return False
+    return True
+
+
+def _key_exists(bucket_name, key):
+    s3 = boto3.resource('s3')
+    try:
+        s3.Object(bucket_name, key).load()
+    except ClientError as e:
+        if e.response['Error']['Code'] == '404':
+            return False
+    return True
+
+
+def _get_file_data(bucket_name, key):
+    s3 = boto3.resource('s3')
+    return s3.Object(bucket_name, key).get()['Body'].read()
+
+
+@mock_s3
+def test_fs_observer_started_event_creates_bucket(observer, sample_run):
+    _id = observer.started_event(**sample_run)
+    run_dir = os.path.join(BASEDIR, str(_id))
+    assert _bucket_exists(bucket_name=BUCKET)
+    assert _key_exists(bucket_name=BUCKET,
+                       key=os.path.join(run_dir, 'cout.txt'))
+    assert _key_exists(bucket_name=BUCKET,
+                       key=os.path.join(run_dir, 'config.json'))
+    assert _key_exists(bucket_name=BUCKET,
+                       key=os.path.join(run_dir, 'run.json'))
+    config = _get_file_data(bucket_name=BUCKET,
+                            key=os.path.join(run_dir, 'config.json'))
+
+    assert json.loads(config.decode('utf-8')) == sample_run['config']
+    run = _get_file_data(bucket_name=BUCKET,
+                         key=os.path.join(run_dir, 'run.json'))
+    assert json.loads(run.decode('utf-8')) == {
+        'experiment': sample_run['ex_info'],
+        'command': sample_run['command'],
+        'host': sample_run['host_info'],
+        'start_time': T1.isoformat(),
+        'heartbeat': None,
+        'meta': sample_run['meta_info'],
+        "resources": [],
+        "artifacts": [],
+        "status": "RUNNING"
+    }
+
+
+@mock_s3
+def test_fs_observer_started_event_increments_run_id(observer, sample_run):
+    _id = observer.started_event(**sample_run)
+    _id2 = observer.started_event(**sample_run)
+    assert _id + 1 == _id2
+
+
+def test_s3_observer_equality():
+    obs_one = S3FileObserver.create(bucket=BUCKET, basedir=BASEDIR)
+    obs_two = S3FileObserver.create(bucket=BUCKET, basedir=BASEDIR)
+    different_basedir = S3FileObserver.create(bucket=BUCKET,
+                                              basedir="another/dir")
+    assert obs_one == obs_two
+    assert obs_one != different_basedir
+
+
+@mock_s3
+def test_raises_error_on_duplicate_id_directory(observer, sample_run):
+    observer.started_event(**sample_run)
```

Comment:
I'm confused by why this would depend on order of test execution; I just tested changing the test name such that it runs either first or last in alphabetical test execution order, and it passed in both cases. I believe that the "create a non failing run and then create a second with the same ID" is what the test is currently doing, so I don't understand your comment well enough to respond to it

Reply:
Must have been a logic error on my side. For the file storage observer, this code would depend on the `started_event` being the first executed in the directory to actually get the id 1. In this case `mock_s3` probably provides a fresh mocked bucket for every test, so never mind. 
Please only remove the `z` from the test name.